### ISSUE TITLE
on cargo publish, check for verified email before compiling

### ISSUE THREAD

**Issue by EverlastingBugstopper** [rust-lang/cargo#8221]
**Describe the problem you are trying to solve**
when running `cargo publish`, the binary is compiled with release mode, and then it checks if your email is verified and you can publish to the crates.io registry.

**Describe the solution you'd like**
i'd like cargo to check if your email is verified and allowed to publish to the crates.io registry before compiling the binary in release mode.

**Comment by metalwarrior665** [comment]
Yes please! I'm now compiling for 4th time and always get some error at the end 👍

**Comment by QiangHeisenberg** [comment]
Is it more reasonable to verify the email address during `cargo login`?
I mean when executing `cargo login xxxxxx`, the system returns a message indicating that the email address has not been verified.

**Comment by EverlastingBugstopper** [comment]
@QiangHeisenberg - it may make sense to add that as a "warning" when you run `cargo login`, but I don't think an unverified email should cause `login` to fail (I think this would be a breaking change). 

However, an unverified email address _already_ makes a `publish` fail, this issue proposes moving that failure earlier in the process (see: before running an expensive compile).

What do you think?

**Comment by QiangHeisenberg** [comment]
I think your suggestion is very reasonable. This is the most prudent approach.

**Comment by epage** [comment]
Could someone help clarify to me which email is being talked about?  Is this for validating that your account on crates.io is setup correctly and you are authorized to publish the package?

**Comment by ehuss** [comment]
Yes, that is my understanding.

**Comment by epage** [comment]
Ok, so this is something that can only be solved by checking with the registry.

**Comment by epage** [comment]
> However, an unverified email address already makes a publish fail, this issue proposes moving that failure earlier in the process (see: before running an expensive compile).

As the motivation for this is to find out about error before slow operations and we sped up those operations in #15783, I'm going to go ahead and close this.  If this was an issue about an error that happens after some packages in a workspace get published but not others, moving the error earlier would be helpful for avoiding a partially published state.  In this case, the failure should be all-or-nothing.

FYI #14941 could possibly speed things up even further.

If there is a reason for us to re-evaluate this, let us know.
